address race condition in LocalClient

mattp- opened the pull request:
### What does this PR do?
The current flow through localclient to Publisher requires waiting for
a response (which contains the jid) before subscribing to the job rets.
There is a race condition in which the publish and response could be
sent before return & subscribe, and there is no reason we can't mint the
jid _before_ publishing and circumvent the issue entirely, so we do
that. The existing master-side code path will still exist for back
compat if a jid is not pre-minted.


### What issues does this PR fix or reference?
tentative race condition in localclient with fast returning calls.

### Tests written?
Yes, added a unit test. to assert jid generation.
**[NOTICE] Bug fixes or features added to Salt require tests.**
Please review the [test documentation](https://docs.saltstack.com/en/latest/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite.


### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

mattp- committed:
address race condition in LocalClient

The current flow through localclient to Publisher requires waiting for
a response (which contains the jid) before subscribing to the job rets.
There is a race condition in which the publish and response could be
sent before return & subscribe, and there is no reason we can't mint the
jid _before_ publishing and circumvent the issue entirely, so we do
that. The existing master-side code path will still exist for back
compat if a jid is not pre-minted.

mattp- committed:
address review comments

mattp- commented on salt/client/__init__.py:
```diff
@@ -1760,6 +1791,8 @@ def pub(self,
             if listen and not self.event.connect_pub(timeout=timeout):
                 raise SaltReqTimeoutError()
+
+            log.error(channel)
```
woops - removed this it was not intended for commit.

mattp- commented on salt/client/__init__.py:
```diff
@@ -1698,6 +1699,35 @@ def _prep_pub(self,
+            jid = self.returners[fstr](nocache=nocache, passed_jid=jid)
+        except (Exception) as exc:
+            # The returner is not present
+            msg = (
```
fixed

mattp- commented on salt/client/__init__.py:
```diff
@@ -1698,6 +1699,35 @@ def _prep_pub(self,
+        try:
+            # Retrieve the jid
+            jid = self.returners[fstr](nocache=nocache, passed_jid=jid)
+        except (Exception) as exc:
```
hmm I'm not sure what to do on this one. since this is passing execution to an arbitrary returner, how do I know what exception class to catch?
I left this as general exception because if it does throw, it will leave jid as '', and send off to master where a second attempt will be made here:
https://github.com/saltstack/salt/blob/develop/salt/master.py#L2209
which seems to only catch KeyError & TypeError. I'm not sure that is desirable either.

s0undt3ch replied:
Ok. Lets leave it then, I also do Not know which in particular to use.

s0undt3ch commented on salt/client/__init__.py:
```diff
@@ -1698,6 +1699,33 @@ def _prep_pub(self,
+            # Retrieve the jid
+            jid = self.returners[fstr](nocache=nocache, passed_jid=jid)
+        except Exception:
+            # The returner is not present
```
I should also have mentioned this on the initial review, don't use string formatting, `{}` in log calls, use string substitution, `log.info('I have a %s to say', 'thing')`

mattp- replied:
fixed this - updated in master.py as well

mattp- committed:
remove format() from log

s0undt3ch commented:
Thanks for putting up with my _review_ :)

mattp- committed:
improve long polling performance / subscription tracking

when dealing with long running jobs via cmd_iter, the heartbeat find_job
calls would continually add new event.subscribe() tag match
subscriptions for the jobs as they come in. This works out to a slow
leak over time when dealing with a job that runs over 12 hours+, and
adds delays when the real job ret finally does come in. This improves
the subscription handling to unsub the previous saltutil.find_job calls
events if set before broadcasting the next.

mattp- committed:
write this more cleanly to appease lint

mattp- commented:
@s0undt3ch @rares-pop @Akm0d apologies for the pre-emptive review, I've added another commit here after encountering an issue earlier today, which tightens up how event subscriptions are handled with long running (12+ hour) jobs. I would appreciate if you can take a look at the latest commit. thanks 👍

s0undt3ch reviewed (commented):
Could we also have a test for the over subscription fix please?
Thanks!

mattp- committed:
add unit tests exercising event subscription logic

mattp- commented:
@s0undt3ch done, added a unit test exercising get_iter_returns event subscriptions to tests.unit.test_client.

mattp- committed:
lint

dwoz commented:
This needs to be rebased to the `master` branch

sagetherage commented:
closing as abandoned
